### DISCUSSION TITLE
Localization questions

### DISCUSSION THREAD

**Discussion by NotroDev** [AvaloniaUI/Avalonia#12668] [Q&A] [answered]
Hello!

I want to add languages to my Avalonia app.
I'm trying to understand this page:
https://docs.avaloniaui.net/docs/next/guides/implementation-guides/localizing

But I have some questions.
1. In the guide, there's this line of code: `Assets.Resources.Culture = new CultureInfo("fil-PH");`
Maybe I'm doing something wrong, but I don't have Assets class.

2. If I have two files: `Resources.resx` and `Resources.pl.resx` it uses `pl` file. Why? Shouldn't `Resources.resx` be a default one? Also, when there is only `Resources.resx`, Localization doesn't work at all, is it intended?

3. Can I use other name than `Resources`? For example, I have a folder `Languages`, can I use file names like `English.resx`?

**Comment by Mrxx99** [comment, marked as answer]
I think Assets here is just the namespace where the Resources.Designer.cs and the .resx files ares

You can name it anyway you want but all resx files should have the same base name and then you need the identifier for the language. So e.g. ANameYouLike.en.resx, you can have one file without the identifier that would be your fallback (ANameYouLike resx)

**Reply by NotroDev** [reply]
Thanks, you answered all my questions! :D
Everything works now.

**Comment by NotroDev** [comment]
Hey, I have one more question.

Is it possible to change the language on runtime?

I'm trying to change the culture, but that does nothing. I can only change language on startup.

`Languages.Language.Culture = CultureInfo.GetCultureInfo("pl-PL");`

**Reply by timunie** [reply]
You can write your own markup extension or use DynamicResources (depending on your App size). https://github.com/timunie/Tims.Avalonia.Samples#localization-sample

**Reply by NotroDev** [reply]
I was thinking about that, but didn't know how to do this. Thanks!

**Comment by NotroDev** [comment]
@timunie 

I'm trying to do this, but it can't find my file.

```xaml
<Application.Resources>
	<ResourceDictionary>
		<ResourceDictionary.MergedDictionaries>
			<ResourceInclude Source="/Languages/en-US.axaml" />
		</ResourceDictionary.MergedDictionaries>
	</ResourceDictionary>
</Application.Resources>
```

![obraz](https://github.com/AvaloniaUI/Avalonia/assets/77281214/c6a1188e-9482-4e79-97a0-ee895f10a4b5)

Files are set to `AvaloniaResource` type.
What's going on?


Your sample is on Avalonia 10, and it doesn't work after updating to 11 without errors, so I can't use it.

**Reply by timunie** [reply]
are you sure it is embedded as AvaloniaResource? Please double-check.

**Reply by NotroDev** [reply]
![obraz](https://github.com/AvaloniaUI/Avalonia/assets/77281214/40a04285-ff2d-4fe8-a0ba-c05d3f1a0901)

Oh, and also, the error:
```
Unable to resolve XAML resource "avares://SkEditorAvalonia/Languages/en-US.axaml" in the "SkEditorAvalonia" assembly. Make sure this file exists and is public. Line 18, position 22.	App.axaml	18
```

**Reply by timunie** [reply]
updated ma samples to match v11

**Reply by NotroDev** [reply]
Ohh, thanks! I'll try again :)

**Reply by NotroDev** [reply]
It seems that this problem was caused because of errors in other XAML files.
It works now. Thanks!

**Comment by henrrymoraes83** [comment]
> Hello!
> 
> I want to add languages to my Avalonia app. I'm trying to understand this page: https://docs.avaloniaui.net/docs/next/guides/implementation-guides/localizing
> 
> But I have some questions.
> 
>     1. In the guide, there's this line of code: `Assets.Resources.Culture = new CultureInfo("fil-PH");`
>        Maybe I'm doing something wrong, but I don't have Assets class.
> 
>     2. If I have two files: `Resources.resx` and `Resources.pl.resx` it uses `pl` file. Why? Shouldn't `Resources.resx` be a default one? Also, when there is only `Resources.resx`, Localization doesn't work at all, is it intended?
> 
>     3. Can I use other name than `Resources`? For example, I have a folder `Languages`, can I use file names like `English.resx`?

I had the same issue. I fixed that adding the lines below inside csproj file.

```
<ItemGroup>
  <AvaloniaResource Include="Assets\**" />
</ItemGroup>

<ItemGroup>
  <EmbeddedResource Update="Assets\Resources.resx">
    <Generator>PublicResXFileCodeGenerator</Generator>
    <LastGenOutput>Resources.Designer.cs</LastGenOutput>
  </EmbeddedResource>
</ItemGroup>

<ItemGroup>
  <Compile Update="Assets\Resources.Designer.cs">
    <DesignTime>True</DesignTime>
    <AutoGen>True</AutoGen>
    <DependentUpon>Resources.resx</DependentUpon>
  </Compile>
</ItemGroup>
```

and edit the Resources.Designer.cs

`global::System.Resources.ResourceManager temp = new global::System.Resources.ResourceManager("YOURPORJECT.Assets.Resources", typeof(Resources).Assembly);`

You can customize the name of file as you wish.

**Reply by NotroDev** [reply]
Thanks! But, is it solving the issue that I described in the original post? I solved it 😅

Now, I want to use XAML + DynamicResource way, but I have this problem:
https://github.com/AvaloniaUI/Avalonia/discussions/12668#discussioncomment-6841198

**Reply by henrrymoraes83** [reply]
I read an article in the past that had a solution similar to what you're looking for:

https://www.codeproject.com/Articles/5317972/Theming-and-Localization-Functionality-for-Multipl

I hope it helps.

**Reply by henrrymoraes83** [reply]
Also there is this one:

https://github.com/sakya/AvaloniaLocalizationExample

**Reply by NotroDev** [reply]
I solved it using timunie's sample, but thanks! :)
